plat: rpi: rpi5: scmi_reset: PCIe support implementation

oleksiimoisieiev opened the pull request:
Basic implementation of the brcm,brcmstb-reset driver inspired by the linux kernel implementation. It is based on the commit-id:

c04af98514c2 (Merge remote-tracking branch 'stable/linux-6.6.y' into rpi-6.6.y, 2024-03-19)

For now the following resets are supported:
- bridge and swinit reset for pcie1
- bridge and swinit reset for pcie2

oleksiimoisieiev committed:
plat: rpi: rpi5: scmi_reset: PCIe support implementation

Basic implementation of the brcm,brcmstb-reset driver inspired by
the linux kernel implementation. It is based on the commit-id:

c04af98514c2 (Merge remote-tracking branch 'stable/linux-6.6.y' into rpi-6.6.y, 2024-03-19)

For now the following resets are supported:
- bridge and swinit reset for pcie1
- bridge and swinit reset for pcie2

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Grygorii Strashko <grygorii_strashko@epam.com>

lorc commented on plat/rpi/rpi5/scmi/scmi_reset.c:
```diff
@@ -7,31 +7,143 @@
+	 * operation is typically 14us for the worst case, build some slack
+	 * here.
+	 */
+	udelay(100);
```
Taking into account that this is ARM TF and you are blocking calling CPU in Secure state, maybe 14us will be sufficient?

GrygiriiS replied:
I think it probably should not be here at all, or should per-reset cfg in rpi5_resets

oleksiimoisieiev replied:
I will try without timeout and with 14 us. Also I will try to inspect rpi4 RM probably it has the same controller.

lorc commented on plat/rpi/rpi5/include/rpi5_scmi_resources.h:
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2024 EPAM Systems Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Seems a bit overkill. As I see, TF-A uses SPDX license identifier

oleksiimoisieiev replied:
fixed

lorc commented on plat/rpi/rpi5/scmi/scmi_reset.c:
```diff
@@ -7,31 +7,143 @@
 			       bool assert_not_deassert __unused)
 {
-	return SCMI_NOT_SUPPORTED;
+	int rc = 0;
```
No need to initialize it.

oleksiimoisieiev replied:
fixed

lorc commented on plat/rpi/rpi5/scmi/scmi_reset.c:
```diff
@@ -7,31 +7,143 @@
+		return SCMI_INVALID_PARAMETERS;
+	}
+
+	rc = (assert_not_deassert) ? brcmstb_reset_assert(scmi_id) :
```
You should remove `__unused` tag from function definition.

oleksiimoisieiev replied:
fixed

lorc commented on plat/rpi/rpi5/scmi/scmi_reset.c:
```diff
@@ -7,31 +7,143 @@
 				  unsigned int scmi_id __unused)
 {
-	return NULL;
+	if (scmi_id >= RPI5_SCMIRST_MAX) {
```
You should remove `__unused` tag from function definition.

oleksiimoisieiev replied:
fixed

lorc commented on plat/rpi/rpi5/scmi/scmi_reset.c:
```diff
@@ -7,31 +7,143 @@
+	 * So if reset was asserted then just do deassert, if deasserted
+	 * then assert/deassert.
+	 */
+	if (scmi_id >= RPI5_SCMIRST_MAX) {
```
You should remove `__unused` tag from function definition.

oleksiimoisieiev replied:
fixed

GrygiriiS commented on plat/rpi/rpi5/scmi/scmi_reset.c:
```diff
@@ -7,31 +7,143 @@
+			return rc;
+		}
+	}
+
```
Here should be a delay. In terms of SCMI RESET_DOMAIN_ATTRIBUTES->latency and it should be configurable per reset and so be part of rpi5_resets[].

oleksiimoisieiev replied:
Current arm-tf version does not support latency attribute

lorc reviewed (approved):
`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

oleksiimoisieiev commented:
@lorc I have updated PR after your Ack. 
I have removed non-used extern:
`extern struct scmi_reset rpi5_resets[RPI5_SCMIRST_MAX];`
from rpi5_scmi_resources.h. Please take a look one more time. Sorry for bothering - found this issue only now....

lorc commented:
It is still fine. You can leave my `Ack`

GrygiriiS commented:
`Reviewed-by: Grygorii Strashko <grygorii_strashko@epam.com>`

oleksiimoisieiev commented:
@lorc Tags were added. Could you please merge
